packages/tools/src/utilities/contours/PointsArray.ts

```diff
@@ -14,152 +17,189 @@ export type PolyDataPointConfiguration = {
  * transferring them amongst systems and is planned to have more methods added
  * for generic manipulation of data.
  */
-export abstract class PointsArray<T> {
+export class PointsArray<T> {
+  /** Allow storage for an index value to indicate where this slice is located */
+  public kIndex: number;
+  /**
+   * Sources data for this array.  Just used for external access, not updated
+   * here.
+   */
+  public sources: PointsArray<T>[];
+
   data: Float32Array;
-  dimensions = 3;
-  public length = 0;
+  _dimensions = 3;
+  _length = 0;
+  _byteSize = 4;
+  growSize = 128;
+  array: ArrayBuffer;
 
   constructor(configuration: PolyDataPointConfiguration = {}) {
-    const { initialSize = 1024, dimensions = 3 } = configuration;
-    this.data = new Float32Array(initialSize * dimensions);
-    this.dimensions = dimensions;
+    const {
+      initialSize = 1024,
+      dimensions = 3,
+      growSize = 128,
+    } = configuration;
+    const itemLength = initialSize * dimensions;
+    this.growSize = growSize;
+    // TODO - use resizeable arrays when they become available in all browsers
+    this.array = new ArrayBuffer(itemLength * this._byteSize);
+    this.data = new Float32Array(this.array);
+    this._dimensions = dimensions;
   }
 
-  protected forEach(func: (value: T, index: number) => void, point: T) {
-    for (let i = 0; i < this.length; i++) {
-      func(this.getPoint(i, point), i);
+  public forEach(func: (value: T, index: number) => void) {
+    for (let i = 0; i < this._length; i++) {
+      func(this.getPoint(i), i);
     }
   }
 
-  abstract getPoint(index: number, point: T): T;
+  public get length() {
+    return this._length;
+  }
+
+  public get dimensions() {
+    return this._dimensions;
+  }
+
+  public get dimensionLength() {
+    return this._length * this._dimensions;
+  }
+
+  public getPoint(index: number): T {
+    if (index < 0) {
```

Comment:
Hmm, is this what the developers expect? If the index is negative, it currently loops around. Perhaps we could consider putting it under a flag and throwing an error if the index is negative. We should also make sure to mention the flag for clarity.

Reply:
It is done this way in about a third of the public APIs for this sort of thing, so I don't think it is unusual.
I'm trying to mostly mimic the array access, which does not throw on negative or out of bounds values, but just returns undefined, so I'd really rather not throw here.